planner/core/planbuilder.go

```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
-			columnName.SetText(column.Name.O)
+		expr, _, err := b.rewrite(ctx, col.GeneratedExpr, mockTablePlan, nil, true)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		expr = expression.BuildCastFunction(b.ctx, expr, colExpr.GetType())
+		genColumnID := model.TableColumnID{TableID: tblInfo.ID, ColumnID: col.ColumnInfo.ID}
+		genExprs[genColumnID] = expr
+	}
 
-			colExpr, _, err := mockTablePlan.findColumn(columnName)
-			if err != nil {
-				return nil, err
+	// Get generated columns.
+	var genCols []*expression.Column
+	pkOffset := -1
+	colsMap := make(map[int64]struct{})
+	schema := expression.NewSchema(make([]*expression.Column, 0, len(idx.Columns))...)
+	for _, idxCol := range idx.Columns {
+		for _, col := range tblInfo.Columns {
+			if idxCol.Name.L == col.Name.L {
+				columns = append(columns, col)
+				tblColumns = append(tblColumns, col)
+				schema.Append(&expression.Column{
+					ColName:  col.Name,
+					UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  &col.FieldType})
+				colsMap[col.ID] = struct{}{}
+				if mysql.HasPriKeyFlag(col.Flag) {
+					pkOffset = len(tblColumns) - 1
+				}
+			}
+			genColumnID := model.TableColumnID{TableID: tblInfo.ID, ColumnID: col.ID}
+			if expr, ok := genExprs[genColumnID]; ok {
+				cols := expression.ExtractColumns(expr)
+				genCols = append(genCols, cols...)
 			}
+		}
+	}
+	tblSchema := schema.Clone()
+	for _, col := range genCols {
+		if _, ok := colsMap[col.ID]; !ok {
+			c := table.FindCol(tbl.Cols(), col.ColName.O)
+			if c != nil {
+				col.Index = len(tblColumns)
+				tblColumns = append(tblColumns, c.ColumnInfo)
+				tblSchema.Append(&expression.Column{
+					ColName:  c.Name,
+					UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  &c.FieldType})
+				colsMap[c.ID] = struct{}{}
+				if mysql.HasPriKeyFlag(c.Flag) {
+					pkOffset = len(tblColumns) - 1
+				}
+			}
+		}
+	}
+	if !tbl.Meta().PKIsHandle || pkOffset == -1 {
+		tblColumns = append(tblColumns, model.NewExtraHandleColInfo())
+		handleCol := &expression.Column{
+			DBName:   dbName,
+			TblName:  tblInfo.Name,
+			ColName:  model.ExtraHandleName,
+			RetType:  types.NewFieldType(mysql.TypeLonglong),
+			UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+			ID:       model.ExtraHandleID,
+		}
+		tblSchema.Append(handleCol)
+		pkOffset = len(tblColumns) - 1
+	}
 
-			expr, _, err := b.rewrite(ctx, column.GeneratedExpr, mockTablePlan, nil, true)
+	is := PhysicalIndexScan{
+		Table:            tblInfo,
+		TableAsName:      &tblInfo.Name,
+		DBName:           dbName,
+		Columns:          columns,
+		Index:            idx,
+		dataSourceSchema: schema,
+		Ranges:           ranger.FullRange(),
+		KeepOrder:        false,
+		GenExprs:         genExprs,
+	}.Init(b.ctx)
+	is.stats = property.NewSimpleStats(0)
+	cop := &copTask{indexPlan: is}
+	// It's double read case.
+	ts := PhysicalTableScan{Columns: tblColumns, Table: is.Table}.Init(b.ctx)
```

Comment:
And could you use a specific function to build PhysicalTableScan, and another for building PhysicalIndexScan? I think that will make the code more readable.